doc/cprover-manual/function-contracts.md

```diff
@@ -0,0 +1,502 @@
+
+```c
+int add_overflow(uint32_t a, uint32_t b, uint32_t* out)
+__CPROVER_requires(out != NULL)
+__CPROVER_ensures(__CPROVER_return_value == SUCCESS || __CPROVER_return_value == FAILURE)
+__CPROVER_ensures((__CPROVER_return_value == SUCCESS) ==> (*out == (a + b)))
+{
+    /* ... */
+}
+```
+**Important.** Developers may call functions inside _requires_ and _ensures_
+clauses to better write larger specifications (e.g., predicates). However, at
+this point CBMC does not enforce such functions to be **pure** (i.e., without
+side effects). This will be supported in future versions.
+
+Finally, the _assigns_ clause allows developers to specify a list of memory
+locations which may be written within a function. While an assigns clause may,
+in general, be interpreted with either _writes_ or _modifies_ semantics, this
+design is based on the former. This means that memory not captured by the
+assigns clause must not be written within the given function scope, even if the
+value(s) therein are not modified.  Assigns clauses currently support simple
+variable types and their pointers, structs, and arrays.  Inductive data
+structures and function pointers are left to future work, as their support would
+require changes to CBMC's memory model. In our example, since we only expect
+that the value pointed by `out` may be modified, we annotate the function using
+`__CPROVER_assigns(*out)`. In contrast, if we expected that the pointer may be
+modified as well, we should annotate the function using
+`__CPROVER_assigns(out)`.
+
+```c
+int add_overflow(uint32_t a, uint32_t b, uint32_t* out)
+__CPROVER_requires(out != NULL)
+__CPROVER_ensures(__CPROVER_return_value == SUCCESS || __CPROVER_return_value == FAILURE)
+__CPROVER_ensures((__CPROVER_return_value == SUCCESS) ==> (*out == (a + b)))
+__CPROVER_assigns(*out)
+{
+    /* ... */
+}
+```
+
+## Semantics
+
+The semantics of a function contract *c* of a given function *f* can be
+understood in two contexts.  First, one may consider how a function contract *c*
+is applied to the function body of *f* in order to determine whether *c* is
+a sound characterization of the behavior of *f* (i.e, enforcement).  Second, one
+may consider how the expressions in *c* are treated when a call to *f* is
+replaced by its contract specification (i.e., replacement), assuming the
+contract specification *c* is a sound characterization of the behavior of *f*.
+To illustrate these two perspectives, consider the following implementation of
+`add_overflow`.
+
+```c
+int add_overflow(uint32_t a, uint32_t b, uint32_t* out)
+__CPROVER_requires(out != NULL)
+__CPROVER_ensures(__CPROVER_return_value == SUCCESS || __CPROVER_return_value == FAILURE)
+__CPROVER_ensures((__CPROVER_return_value == SUCCESS) ==> (*out == (a + b)))
+__CPROVER_assigns(*out)
+{
+    const uint64_t result = ((uint64_t) a) + ((uint64_t) b);
+    if (result > UINT32_MAX) return FAILURE;
+    *out = (uint32_t) result;
+    return SUCCESS;
+}
+```
+### Enforcement
```

Comment:
Comment: not clear how replacement vs enforcement of contracts might be different or what the use cases are, could be helpful to have that spelled out before diving into the how-to for each.

Reply:
I've broken down this section for each clause individually, so we can explain how both processes apply to each case. I have also included in the first running example (see Overview section) the steps to run both procedures and what their goals are.